models/schema/v1/workload/secret/secret.k

```diff
@@ -37,197 +28,29 @@ schema BasicAuthSecret(SecretBase):
-    data: {str,str}, default is Undefined, optional.
-        Data contains token info if user already known. The key name must be token.
-
-    Examples
-    --------
-    import models.schema.v1.workload.secret as sec
-
-    token = sec.TokenSecret {
-        type: "token"
-        data: {
-            # optional
-            "token": ""
-        }
-    }
-    """
-
-    # The token secret type.
-    type:                       "token"
+    # Types of secrets available to use.
+    type:                       "basic" | "token" | "opaque" | "certificate" | "external"
 
     # Params defines extra parameters used to customize secret handling.
     params?:                    {str:str}
 
-    # Data contains token info if user already known.
+    # Data defines the keys and data that will be used by secret.
     data?:                      {str:str}
 
-    check:
-        all k in data {
-            k in ["token"]
-        } if data, "a valid secret data key name must be token for TokenSecret"
-
-schema OpaqueSecret(SecretBase):
-    """ OpaqueSecret is used for arbitrary key value pairs. This type of secrets are
-    best used for allowing a user to input sensitive data at runtime.
-
-    Attributes
-    ----------
-    type: str, default is opaque, required.
-        The actual type of secret.
-    data: {str,str}, default is Undefined, optional.
-        Data contains arbitrary key value pairs.
-
-    Examples
-    --------
-    import models.schema.v1.workload.secret as sec
-
-    opaqueSec = sec.OpaqueSecret {
-        type: "opaque"
-    }
-    """
-
-    # The opaque secret type.
-    type:                       "opaque"
-
-    # Data contains arbitrary key value pairs.
-    data?:                      {str:str}
+    # If immutable set to true, ensures that data stored in the Secret cannot be updated.
+    immutable?:                 bool
 
     check:
         not data or all k in data {
             regex.match(k, r"[A-Za-z0-9_.-]*")
         }, "a valid secret data key must consist of alphanumeric characters, '-', '_' or '.'"
-
-schema TLSSecret(SecretBase):
-    """ TLSSecret is used for storing a certificate and its associated key that
-    are typically used for TLS.
-
-    Attributes
-    ----------
-    type: str, default is tls, required.
-        The actual type of secret.
-    data: {str,str}, default is Undefined, required.
-        Data contains certificate and its associated key.
-
-    Examples
-    --------
-    import models.schema.v1.workload.secret as sec
-
-    tlsSecret = sec.TLSSecret {
-        type: "tls"
-        data: {
-            # must be provided
-            "tls.crt": ""
-            "tls.key": ""
-        }
-    }
-    """
-
-    # The tls secret type.
-    type:                       "tls"
-
-    # Data contains certificate and its associated key.
-    data:                       {str:str}
-
-    check:
+        all k in data {
```

Comment:
I would suggest using union types to implement different types of Secrets, which is very suitable for extension. Here is a code implementation.
```python
schema SecretBase:
    params?:  {str:str}
    immutable?:                 bool

schema Basic(SecretBase):
    username: str
    password: str

schema Token(SecretBase):
    token: str

schema Certificate(SecretBase):
    "tls.crt": str
    "tls.key": str

type Secret = Basic | Token | Certificate

schema Config:
    secret?: Secret
```

Reply:
Of course, if you still use the combination of type and data fields, I suggest defining a dictionary to maintain it.
```python
SECRET_TYPE_DATA_MAPPING = {
    basic = ["username", "password"]
    token = ["token"]
}

schema Secret:
    type: str
    data?: {str:str} = {}

    check:
          type in SECRET_TYPE_DATA_MAPPING, "type must be one of ${list(SECRET_TYPE_DATA_MAPPING)}"
          all k in data {
                k in SECRET_TYPE_DATA_MAPPING[type]
          }, "a valid secret data key name must be one of ${SECRET_TYPE_DATA_MAPPING[type]} for ${type} type secret, got ${list(data)}"
```